hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecord.java

```diff
@@ -282,6 +287,47 @@ public void checkState() {
     }
   }
 
+  protected abstract void writeRecordPayload(T payload, Kryo kryo, Output output);
+
+  protected abstract T readRecordPayload(Kryo kryo, Input input);
+
+  /**
+   * NOTE: This method is declared final to make sure there's no polymorphism and therefore
+   *       JIT compiler could perform more aggressive optimizations
+   */
+  @Override
+  public final void write(Kryo kryo, Output output) {
+    Serializer recLocSerializer = kryo.getSerializer(HoodieRecordLocation.class);
+
+    kryo.writeObjectOrNull(output, key, HoodieKey.class);
+    kryo.writeObjectOrNull(output, currentLocation, recLocSerializer);
+    kryo.writeObjectOrNull(output, newLocation, recLocSerializer);
```

Comment:
How about the performance gains for records SE/DE after this change ?

Reply:
It's less about performance gains in the `HoodieRecord` per se and more about fixing ser/de for `HoodieSparkRecord`

 - Previous PR actually removed mandatory conversion to `UnsafeRow` w/in the `HoodieSparkRecord` (for performance reasons)
 - Conversion to `UnsafeRow` is still required though upon ser/de (for shuffling). This change is to handle that.